webapp/db/db_schema.py

```diff
@@ -18,63 +18,110 @@ class DbClient(Base):
+    initiated = Column(DateTime, default=datetime.now(), index=True)
     expires   = Column(DateTime, default=datetime.now())
 
-    def __init__(self, client_id, id_secret, challenge_secret):
+    challenge: Mapped["DbClientAuthChallenge"] = relationship(cascade="delete")
+
+    def __init__(self, client_id):
         self.id = _gen_uuid()
         self.client_id = client_id
-        self.id_secret = id_secret
         self.auth_state = DbAuthState.PENDING
-        self.challenge_secret = challenge_secret
         self.initiated = datetime.now()
 
     def activate(self, expires):
-        self.auth_state = DbAuthState.ACTIVE
+        self.auth_state = DbAuthState.SUCCESSFUL
         self.expires = expires
-        self.challenge_secret = None
-        self.id_secret = None
     
     def fail(self):
         self.auth_state = DbAuthState.FAILED
-        self.challenge_secret = None
-        self.id_secret = None
 
-class DbClientRepoAccess(Base):
+class DbClientAuthChallenge(Base):
+    __tablename__ = "client_auth_challenges"
+
+    auth_event_id: Mapped[String] = mapped_column(ForeignKey('client_auth_sessions.id'), primary_key=True)
+
+    id_secret = Column(String, unique=True)
+    challenge_secret = Column(String, unique=True)
+
+    def __init__(self, auth_event_id, id_secret, challenge_secret):
+        self.auth_event_id = auth_event_id
+        self.id_secret = id_secret
+        self.challenge_secret = challenge_secret
+
+
+class DbGitCommit(Base):
+    """ Table for tracking the Git Commits that have been the HEAD """
+
+    __tablename__ = "repo_commits"
+
+    commit_hash = Column(String, primary_key=True)
+
+    commit_time = Column(DateTime)
+
+    sync_time = Column(DateTime)
+
+    def __init__(self, commit_hash, commit_time):
+        self.commit_hash = commit_hash
+        self.commit_time = commit_time
+        self.sync_time = datetime.now()
+
+class DbClientCommitAccess(Base):
     """ Table for tracking the latest version of the git repo accessed by a client """
-    __tablename__ = "client_git_access"
+    __tablename__ = "client_commit_access"
     
     id = Column(String, primary_key=True, default = _gen_uuid)
     client_id: Mapped[String] = mapped_column(ForeignKey('client.id'))
     
-    commit_hash = Column(String)
+    commit_hash: Mapped[String] = mapped_column(ForeignKey('repo_commits.commit_hash'))
     access_time = Column(DateTime)
 
-    def __init__(self, client_id, commit_hash):
+    def __init__(self, client_id, commit_hash, access_time = None):
         self.id = _gen_uuid()
         self.client_id = client_id
         self.commit_hash = commit_hash
+        self.access_time = access_time
+
+
+class DbClientStateView(Base):
+    """
+    Db Model class for the non-manifested view that displays the latest state for each client
+    prior to a given timestamp (see client_state_report.py)
+    # TODO figure out a way to explicitly mark this as a view - SQLAlchemy currently creates
```

Comment:
Are there any implications from this? How does it get resolved into a view?

Reply:
@iross currently it doesn't, the process for making a view in SQLAlchemy is [well-documented](https://github.com/sqlalchemy/sqlalchemy/wiki/Views) but requires quite a bit of custom code. Currently this just creates an empty table that's never inserted to/selected from. The model class is only used to automatically convert the result of the big multi-table join query in [client_state_report.py](https://github.com/CHTC/gm-file-server/pull/6/files#diff-c26c6e63532221890d9b063ccaf85521f85ced92e6a53753c404b340aa0c6712R9) to Python objects.